File: neetcode150-py/src/neetcode150_py/arrays_and_hashing/q1_two_sum.py
from typing import List, Dict, Optional


class Solution:
    def twoSum(self, nums: List[int], target: int) -> Optional[List[int]]:
        """
        T(n) = O(n)
        S(n) = O(n)
        """
        val2idx: Dict[int, int] = {}
        for i in range(len(nums)):
            want = target - nums[i]
            if want in val2idx:
                return [i, val2idx[want]]
            val2idx[nums[i]] = i
        return None

    # def twoSum(self, nums: List[int], target: int) -> List[int]:
    #     """
    #     T(n) = O(n^2)
    #     S(n) = O(1)
    #     """
    #     for i in range(len(nums)):
    #         for j in range(i + 1, len(nums)):
    #             if nums[i] + nums[j] != target:
    #                 continue
    #             return [i, j]
    #     return []

File: neetcode150-rs/src/arrays_and_hashing/q1_two_sum.rs
use std::collections::HashMap;

pub struct Solution;

impl Solution {
    /// - T(n) = O(n)
    /// - S(n) = O(n)
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map: HashMap<i32, usize> = HashMap::new();

        for (i, &num) in nums.iter().enumerate() {
            // Check if map contains target - num, if yes, return, otherwise, insert value
            let diff = target - num;
            if let Some(&j) = map.get(&diff) {
                return vec![i as i32, j as i32];
            }
            map.insert(num, i);
        }
        unreachable!()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use itertools::sorted;

    #[test]
    fn test_two_sum() {
        assert_eq!(
            sorted(Solution::two_sum(vec![2, 7, 11, 15], 9)).collect::<Vec<i32>>(),
            vec![0, 1]
        );
        assert_eq!(
            sorted(Solution::two_sum(vec![3, 2, 4], 6)).collect::<Vec<i32>>(),
            vec![1, 2]
        );
        assert_eq!(
            sorted(Solution::two_sum(vec![3, 3], 6)).collect::<Vec<i32>>(),
            vec![0, 1]
        );
    }
}
